### ISSUE TITLE
`borrow_deref_ref` suggestion breaks code

### ISSUE THREAD

**Issue by Rimpampa** [rust-lang/rust-clippy#11346]
### Summary

When there is a method (not a freestanding function) that takes `self: &mut &Self` (or simply `&mut self` when `Self` is `&T`) and such a method is called on an immutable variable of the `&Self` type, a reborrow is required so that _a mutable reference to the copy of the original reference_ is taken.

The problem is that clippy flags such a reborrow as needless, when in reality, when removed, it causes a compilation error because it is not possible to take a mutable reference to an immutable variable.

### Lint Name

borrow_deref_ref

### Reproducer

This is the most minimal example I managed to write that reproduces the problem:

```rust
struct Struct;

impl Struct {
    fn foo(self: &mut &Self) {}
}

trait Trait {
   fn bar(&mut self) {}
}

impl Trait for &Struct {}

fn main() {
    let s = &Struct;

    (&*s).foo(); // warning: deref on an immutable reference
//  ^^^^^ help: if you would like to reborrow, try removing `&*`: `s`

    (&*s).bar(); // warning: deref on an immutable reference
//  ^^^^^ help: if you would like to reborrow, try removing `&*`: `s`
}
```

---

Note that if we remove the need to use the dot operator, the lint doesn't show up:
```rust
struct Struct;

fn foo(_: &mut &Struct) {}

fn main() {
    let s = &Struct;
    foo(&mut &*s) // no warning (reborrow still required)
}
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
